enhancements/kubelet/virtualization-higher-workload-density.md

```diff
@@ -0,0 +1,570 @@
+virtual address space, this is required in order to reduce the
+likelihood for workloads to OOM significantly.
+
+Besides _increasing_ the virtual address space, there are two
+mechanisms which help to reduce the memory footprint of a workload.
+
+They are not being discussed in depth here, but are mentioned, as they
+help to reduce or avoid memory pressure, which in turn leads less usage
+of swap.
+
+The mechanisms are:
+* _Free Page Reporting ([FPR])_ - A mechanism to free memory on the
+  hypervisor side if it is not used by the guest anymore.
+* _Kernel Samepage Merging ([KSM])_ - A mechanism to merge/deduplicate
+  memory pages with the same content. This is particularly helpful
+  in situations where similar guests are running on a hypervisor.
+  It has however some security implications.
+
+##### Differences between Technology Preview vs GA
+
+|                              | TP                  | GA                  |
+|------------------------------|---------------------|---------------------|
+| SWAP Provisioning            | MachineConfig       | MachineConfig       |
+| SWAP Eligibility             | burstable QoS pods  | burstable QoS pods  |
+| Node service protection      | Yes                 | Yes                 |
+| I/O saturation protection    | Yes                 | Yes                 |
+| Critical workload protection | No                  | Yes                 |
+| Memory pressure handling     | Memory based        | Memory & Swap based |
+
+### Risks and Mitigations
+
+#### Phase 1
+
+| Risk                                       | Mitigation                                                                                               |
+|--------------------------------------------|----------------------------------------------------------------------------------------------------------|
+| Miss details and introduce instability     | * Adjust overcommit ratio <br/> * Tweak eviction thresholds <br/> * Use de-scheduler to balance the load |
+
+#### Phase 2
+
+Swap is handled by upstream Kubernetes.
+
+| Risk                                                      | Mitigation                                        |
+|-----------------------------------------------------------|---------------------------------------------------|
+| Swap-based evictions are based on API-initiated evictions | Also rely on kubelet-level memory-based evictions |
+
+#### Phase 3
+
+Upstream Kubernetes handles both swap and evictions.
+Swap provision handled by OpenShift.
+
+### Drawbacks
+
+The major drawback and risk of the [WASP Agent] approach in phase 1 is
+due to the lack of integration with Kubernetes. It's prone to
+regressions due to changes in Kubernetes.
+
+Thus phase 2 is critical in order to eliminate those risks and
+drawbacks.
+
+## Open Questions [optional]
+
+None.
+
+## Test Plan
+
+Add e2e tests for the WASP agent repository for regression testing against
+OpenShift.
```

Comment:
I think that we should include here a bit more details about how we are (already) testing it. Most importantly: configure 200% over-commitment, fill up the cluster with dormant VMs and verify that the cluster is responsive and survives upgrade.